File: thermometer.rb
#!/usr/bin/env ruby

=begin
Copyright © 2013 Baiki <dot_baiki@yahoo.com>
This work is free. You can redistribute it and/or modify it under the
terms of the Do What The Fuck You Want To Public License, Version 2,
as published by Sam Hocevar. See the COPYING file for more details.
=end

#%w[sinatra thin data_mapper dm-sqlite-adapter bcrypt].each { |g| require g }
%w[sinatra thin].each { |g| require g }

SOFTWARE_NAME     = 'Thermometer'
SOFTWARE_VERSION  = 'v0.18'

configure do
  set :bind, '0.0.0.0'
  set :port, 1234
  enable :logging
#  enable :sessions
  use Rack::Session::Cookie, :key => 'thermometer.session',
                             :path => '/',
                             :expire_after => 900,
                             :secret => 'hH76Dvj90Lxb157aaPUw'
end

helpers do
  def link(name)
    case name
    when :home then '/'
    when :thermometer then '/thermometer'
    when :chart then '/chart'
    when :logout then '/logout'
    else "/"
    end
  end

  def get_readings
    actual_file = Dir.glob("data/messwerte*").max_by {|f| File.mtime(f)}
    @temp_date, @temp_time, @temp_celcius = `tail -n 1 #{actual_file}`.rstrip.split(',')
    @cpu_temp = (`cat /sys/class/thermal/thermal_zone0/temp`.to_f / 1000).round(1)
  end

  def get_statistics
    statistics = IO.read('data/gruff-statistics')
    @total_readings = statistics.chomp
  end

  def protected!
    return if authorized?
    headers['WWW-Authenticate'] = 'Basic realm="Protected Access"'
    halt 401, "Ouch... not authorized\n"
  end

  def authorized?
    @auth ||=  Rack::Auth::Basic::Request.new(request.env)
    @auth.provided? and @auth.basic? and @auth.credentials and @auth.credentials == ['1', '1']
  end
end

before do
end

after do
end

get '/' do
  redirect link(:thermometer)
end

get '/thermometer' do
  get_readings
  get_statistics
  erb :thermometer
end

get '/chart' do
  get_readings
  get_statistics
  erb :chart
end

get '/logout' do
  get_readings
  session.clear
  erb :logout
end

get '/admin' do
  protected!
  get_readings
  erb :admin
end

not_found do
  "Ouch... not found."
end

error do
  "Ouch... error."
end

File: thermometer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

'''
Copyright © 2013 Your Name <dot_baiki@yahoo.com>
This work is free. You can redistribute it and/or modify it under the
terms of the Do What The Fuck You Want To Public License, Version 2,
as published by Sam Hocevar. See the COPYING file for more details.
'''

import os, glob, time, signal, sys

version      = 0.54
new_temp_c   = 0.0
old_temp_c   = 0.0
count        = 0
ausgabedatei = 'data/messwerte_' + str(time.time()) + '.csv'

try:
    datei = open(ausgabedatei, 'a')
except:
    print 'File access failed. Programm terminated.'
    sys.exit(0)

os.system('sudo modprobe w1-gpio')
os.system('sudo modprobe w1-therm')

base_dir      = '/sys/bus/w1/devices/'
device_folder = glob.glob(base_dir + '28*')[0]
device_file   = device_folder + '/w1_slave'

def read_temp_raw():
    f = open(device_file, 'r')
    lines = f.readlines()
    f.close()
    return lines

def read_temp():
    lines = read_temp_raw()
    while lines[0].strip()[-3:] != 'YES':
        time.sleep(0.2)
        lines = read_temp_raw()
    equals_pos = lines[1].find('t=')
    if equals_pos != -1:
        temp_string = lines[1][equals_pos+2:]
        temp_c = float(temp_string) / 1000.0
        return temp_c

def signal_handler(signal, frame):
    print '\nCTRL+C received. Programm terminated.'
    datei.close()
    sys.exit(0)

signal.signal(signal.SIGINT, signal_handler)

os.system('clear')
print "Baiki's Thermometer v" + str(version)
print "(press CTRL+C to quit)\n"

while True:
    new_temp_c = round(read_temp(), 1)
    if new_temp_c != old_temp_c:
        loctime = time.localtime()
        j, m, t, std, min, sec = loctime[0:6]
        datum = "{0:04d}.{1:02d}.{2:02d}".format(j, m, t)
        zeit = "{0:02d}:{1:02d}:{2:02d}".format(std, min, sec)
        print '(' + datum, zeit + ') ' + str(new_temp_c) + '°C'
        datei.write(datum + ',' + zeit + ',' + str(new_temp_c) + "\n")
        count += 1
        if count >= 5:
            datei.close()
            datei = open(ausgabedatei, 'a')
            count = 0
        old_temp_c = new_temp_c
        time.sleep(10)
